cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -534,47 +546,126 @@ public Void call() throws Exception {
     });
   }
 
-  private void setInputDatasetIfNeeded(Job job) throws IOException {
-    InputFormatProvider provider = context.getInputFormatProvider();
+  // Multiple mappers could be defined, so we check that their output types are consistent.
+  // Return the TypeToken for one of the mappers (doesn't matter which one, since we check that all of their output
+  // key/value types are consistent).
+  // Returns null if their are no inputs configured.
+  @Nullable
+  private TypeToken<?> checkMapperOutputTypes(Job job) throws ClassNotFoundException {
+    int numInputs = context.getMapperInputs().size();
+    if (numInputs == 0) {
+      // if there are no inputs configured through CDAP APIs, nothing to check for consistency
+      return null;
+    }
 
-    // If input format is not set during beforeSubmit, use the one from the spec if it exists.
-    if (provider == null && specification.getInputDataSet() != null) {
-      context.setInput(specification.getInputDataSet());
-      provider = context.getInputFormatProvider();
+    // get all the mappers configured through our APIs
+    List<Class<? extends Mapper>> mapperClasses = new ArrayList<>();
+    for (MapperInput mapperInput : context.getMapperInputs().values()) {
+      Class<? extends Mapper> mapper = mapperInput.getMapper();
+      if (mapper != null) {
+        mapperClasses.add(mapper);
+      }
     }
 
-    if (provider != null) {
-      Configuration jobConf = job.getConfiguration();
-      ConfigurationUtil.setAll(provider.getInputFormatConfiguration(), jobConf);
-      jobConf.set(Job.INPUT_FORMAT_CLASS_ATTR, provider.getInputFormatClassName());
+    if (mapperClasses.size() < numInputs) {
+      // means one of the inputs is going to default to use the mapper assigned on the job,
+      // so also check the job's mapper class
+      if (!isProgrammaticConfig(job.getConfiguration(), Job.MAP_CLASS_ATTR)) {
```

Comment:
Shouldn't this be an error already if any of the configured input doesn't have a mapper class and also the mapper class was not set in the job directly?